module/Finna/src/Finna/Record/IIIF/IIIFManifestGenerator.php

```diff
@@ -0,0 +1,181 @@
+use VuFind\View\Helper\Root\RecordLinker;
+
+/**
+ * IIIF manifest generator service
+ *
+ * Only intended for internal use as a compatibility layer. With this we can use
+ * Tify to show non-IIIF images and image sets.
+ *
+ * @category VuFind
+ * @package  Content
+ * @author   Ronja Koistinen <ronja.koistinen@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class IIIFManifestGenerator implements
+    \VuFindHttp\HttpServiceAwareInterface
+{
+    use \VuFindHttp\HttpServiceAwareTrait;
+
+    /**
+     * Constructor.
+     *
+     * @param Url          $url          URL helper
+     * @param ServerUrl    $serverUrl    Server URL helper
+     * @param RecordLinker $recordLinker RecordLinker helper
+     *                                   For getting the URL of the record
+     *                                   action constructing this class
+     */
+    public function __construct(
+        protected Url $url,
+        protected ServerUrl $serverUrl,
+        protected RecordLinker $recordLinker,
+    ) {
+    }
+
+    /**
+     * Generate IIIF presentation manifest (version 3)
+     *
+     * @param RecordDriver $driver Record driver
+     *
+     * @return array|null
+     */
+    public function generate(RecordDriver $driver): array|null
+    {
+        $images = $driver->tryMethod('getAllImages');
+        if (!$images) {
+            return null;
+        }
+
+        $recordId = $driver->getUniqueID();
+        $manifestId = ($this->serverUrl)(
+            $this->recordLinker->getActionUrl(
+                $driver,
+                'IIIFManifest',
+                options: ['force_canonical' => true]
+            )
+        );
+
+        $manifest = [
+            '@context' => 'http://iiif.io/api/presentation/3/context.json',
+            'id' => $manifestId,
+            'type' => 'Manifest',
+            'thumbnail' => [],
+            'metadata' => [],
+            'items' => [],
+        ];
+
+        foreach ($images as $idx => $image) {
+            $canvasItem = [
+                'id' => "$manifestId/$idx",
+                'type' => 'Canvas',
+                'items' => [],
+            ];
+            foreach (['large', 'medium', 'small'] as $size) {
+                if (isset($image['urls'][$size])) {
+                    $canvasItem['items'][] =
+                        $this->createAnnotationPage(
+                            $recordId,
+                            $idx,
+                            $size,
+                            $driver,
+                            $manifestId
+                        );
+                    break; // only take the largest $size
+                }
+            }
+            $manifest['items'][] = $canvasItem;
+        }
+
+        if (empty($manifest['items'])) {
+            return null;
+        } else {
+            return $manifest;
+        }
+    }
```

Comment:
Vaiko näin:

```php
return empty($manifest['items']) ? null : $manifest
```

Reply:
Kun `$manifest['items']` on aina array, niin tämä on yksinkertaisin:
```suggestion
        return $manifest['items'] ? $manifest : null;
    }
```

Reply:
Toinen vaihtoehto on koota `$items`-array ennen kuin kasaat manifestin. Jos ei löydy mitään, niin voit palauttaa null:n jo ennen kuin haet edes $recordTitle:a tai alat kasaamaan `$manifest`:ia.